### ISSUE TITLE
[Question] Documentation best practices

### ISSUE THREAD

**Issue by emosher** [ngbp/ngbp#303]
I'm using ngbp and I was wondering what ngbp creators/users think about documenting their code. Given that ngbp is an "opinionated" build system, I was wondering what others thought. :smile: 

I typically gravitate towards YUIdoc, so the following examples assume that. In non-ngbp projects, the solutions was dictated by the `app.directives`, `app.services`, etc module system. I could then document a directive like so:

``` javascript
/**
 * myDirective does some DOM stuff
 *
 * @namespace directives
 * @class myDirective
 * @constructor
 */
```

With ngbp, I could certainly make additional modules off of the component modules like `angular.module('myApp.componentOne.directives', [])`. The problem I see with this is that I'm creating additional modules for no reason. With ngbp `componentOne` is typically no more than 1-2 JS files and perhaps a template. A directive module seems like overkill.

I've thought about a few other options including YUIdoc's `@extends` tag, other documentation engines, and plugins/forks of YUIdoc, JSdoc, etc.  So far I haven't been able to find a solution that results in the docs actually being useful.

Would appreciate thoughts/help.  Thanks for making ngbp!

**Comment by PaulL1** [comment]
Have you considered ngDoc?

**Comment by joshdmiller** [comment]
@emosher There are a number of preferences that enter into documentation and I would be hard-pressed to name the one I thought was the "best practice". Some like to generate API documentation from C-like source comments, such as through YUIdoc or ngDoc, and others like to [keep documentation separate](https://github.com/millermedeiros/mdoc). I tend to fall on the latter side as I find inline comments get in the way of reading source code, but I'm probably in the minority.

I'm not familiar enough with YUIdoc to provide much guidance here, except to say that your instinct is probably right and creating additional modules just for the sake of ordered documentation is overkill. As @PaulL1 mentioned, ngDoc can probably help with this as it contains [the `@ngdoc` comment directive](https://github.com/angular/angular.js/wiki/Writing-AngularJS-Documentation#the-ngdoc-directive) that allows you to specify what type of thing the comment is documenting.

I hope others will chime in too.

**Comment by emosher** [comment]
Thanks for the replies. I have looked at ngDoc, [nag-doc](https://github.com/nucleus-angular/nag-doc), [Dgeni](https://github.com/angular/dgeni) and others. I'm mainly curious to see how others that are using ngbp are creating their documentation. I'm particularly interested in why you might choose one documentation engine over another because there are several that are tailored to Angular, as well as the vanilla JS engines like YUIdoc and JSDoc.  

Part of the issue I see is aligning an Angular frontend (ngbp) with other JS parts of a project. For example, a Node.js backend or a (non-Angular) JS library being developed as part of a larger ecosystem.  Hoping to hear more opinions on this.

@joshdmiller I appreciate the philosophy behind mdoc and the blog post by the author. I understand the thought process behind it, but I can't say that I agree. Thanks for the response!

**Comment by PaulL1** [comment]
I think I agree with Josh in principle, inline comments have some bad aspects.

I've been contributing a little to the ng-grid (about to be ui-grid) project, and it uses ngdoc.  I spent time in the ngdoc fixing it, and in the last few days I decided to move my application to ngdoc.  

The reasons for this were:
1. Yes, it makes the code harder to read.  But actually, I'd put comments in the code anyway, it's not like I'd move to zero comments
2.  One potential future for my app as part of me becoming an internet zillionaire&trade; is that I sell it to someone else to maintain (buyout or whatever).  In that situation the someone would want documentation.  Whilst ngdoc may make the code harder to read, it's a box to tick to show you had "good practice"
3.  I kinda like the documentation site that it creates.  I know that I'll never use it....but it's flashy and nice
4.  When I went through documenting the code, I found a bunch of refactoring.  Use that as a counter-example, I'm not sure that documenting or not documenting finds things, it's just the bit where you go back and look at the code again that lets you find things.
5. ngdoc has some annoyances, like no way to easily put properties with data types on objects, can't have the same comment apply to multiple objects (I'd like to just go @propertyOf twice and have it turn up in both places), but I can work around them.  It's good enough, and it's the angular way so it'll no doubt grow with angular

In short, whilst I see it as making my code noisy, I think it's a necessary evil and it's not too much more verbose than what I was otherwise doing.

I've done the gruntfile changes for ngdoc if you're interested.  For me, it builds to a doc/ directory in my project, and uses templates etc from dev_tools/doc.  

I installed ngdoc with:

```
  npm install grunt-ngdocs --save-dev
```

I added into my gruntfile:

``` javascript
  grunt.loadNpmTasks('grunt-ngdocs');
```

near the top of the file, and straight after the protractor section (around line 450 - don't go further down into the delta task by accident):

``` javascript
    ngdocs: {
      options: {
        dest: 'docs',
        startPage: '/api',
        html5Mode: false,
        title: "your title",
        styles: [
          'dev_tools/doc/css/prettify.css',
          'dev_tools/doc/css/bootstrap-flatly.css'
        ],
        navTemplate: 'dev_tools/doc/templates/nav.html'
      },
      api: {
        src: ['src/**/*.js', '!src/**/*.scenario.js'],
        title: 'Documentation'
      },
      functional_test: {
        src: ['src/**/*.scenario.js'],
        title: 'Functional Test'
      }
    },     
```

The functional test block was actually the main bit I wanted to document, someone else will be building the end to end functional testing based on the library of functions I wrote and some example scripts, so I put that into a separate functional testing section.  It's uglier because there are no modules in the functional tests, and ngdoc looks to lean on angular modules for some of it's structure.  But it works.

The templates etc can be copied from the ng-grid github, and that's enough to get started.

I haven't integrated it into my workflow - I'm just building directly grunt ngdocs, rather than having them built with every iteration.  I should put into the watch, they only take a half second to run, but I haven't.

At some point I'll do it properly, but I'd target warlock for that rather than ngbp probably.

Also, I spent a while trying to get it to work and failing, and then finally noticed how important it is that each comment start with /**.  Probably everyone but me already knew that, but just in case....
